FNX-14633 ⁃ For #162 - Add OSS build flavor.

gilbsgilbs opened the pull request:
This PR adds an OSS build flavor allowing to build fenix without proprietary components, namely:

- play-services-ads-identifier
- play-services-oss-licenses (the plugin is FOSS, but not the library, so it was replaced with [AboutLibraries](https://github.com/mikepenz/AboutLibraries))
- lib-push-firebase (which isn't closed-source by itself yet bundles proprietary libraries such as firebase-core and play-services)
- leanplum (which isn't closed-source by itself but includes dependencies to firebase and gms)

I didn't do the work for other SDKs that are open-source but promote closed-source tracking services (such as Adjust), as it is not strictly necessary to release fenix on F-Droid as far as I know.

Build the oss / non-oss flavor:
```
./gradlew assembleOssDebug assembleNonOssDebug
```

Simple check for proprietary dependencies:
```console
$ dexdump app/build/outputs/apk/nonOss/debug/app-nonOss-arm64-v8a-debug.apk  | grep -E '(firebase|gms)' | wc -l
21121
$ dexdump app/build/outputs/apk/oss/debug/app-oss-arm64-v8a-debug.apk  | grep -E '(firebase|gms)' | wc -l      
0
```

So I guess it works.

I think this PR even closes #162 as the work of making fenix available on F-Droid will mostly be political after that.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not - mostly consists in stubbing existing implementations
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not - no UI changes
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product. - no UI changes

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

ping @pocmo

NotWoods commented on app/src/free/java/org/mozilla/fenix/components/metrics/LeanplumMetricsServiceImpl.kt:
```diff
@@ -0,0 +1,21 @@
+import android.net.Uri
+import java.util.UUID
+
+class LeanplumMetricsServiceImpl(
```
LeanplumMetricsService is used for two cases, both of which already have interfaces. Rather than making a dummy metrics service, let's configure those two locations.

First one is easy, the service is used in MetricController.create and discarded if the Telemetry flag is set to false. No change needed.

Second location is for the DeepLinkIntentProcessor. It wants a DeepLinkVerifier instance. You could pass in a dummy version that always returns false based on build flags.

NotWoods replied:
Sorry, I think I was getting confused with BuildConfig flags and build variants. I thought this could be accomplished with an if statement. 

Are we allowed to use any form of telemetry in the F-Droid version? If not, we may want to replace the entire MetricController. We have a dummy `DebugMetricController` already that could be used.

gilbsgilbs replied:
> Are we allowed to use any form of telemetry in the F-Droid version?

I'm pretty sure that yes, telemetry is allowed as long as the frameworks being used are completely FLOSS (so Glean and Adjust should be OK, not firebase because of the dependency to gms). You can even display ads under the same conditions.

However, telemetry, especially when enabled by default like it is in Fenix, is considered an "anti-feature". This means there will be a warning prior downloading the app on the store. Such warning is already displayed for Fennec, so that's not too bad I guess.

NotWoods commented on app/src/free/java/org/mozilla/fenix/settings/about/OssLicensesMenuImpl.kt:
```diff
@@ -0,0 +1,15 @@
+
+class OssLicensesMenuImpl : OssLicensesMenu {
+    override fun getIntent(context: Context?, title: String): Intent {
+        // This menu shouldn't displayed on the free variant anyways.
```
Why not? AFAIK open source libraries should be fine to leave in, and we'll need to credit them to use them in the app.

NotWoods replied:
I'm still a bit confused here. Why can't we use the Google library? As far as I can tell its open source and just figures out the play services libraries we use, rather than depending on them. 

If we need to change the license implementation I think it would be nice to have it split into a separate PR so its easier to review.

gilbsgilbs replied:
Actually it **is** confusing. The gradle plugin that lists the dependencies is open-source and under Apache 2 license, but the library that is shipped with the APK is proprietary and, as far as I can tell, closed-source. Morever, the library has dependencies to gms (which you can easily see with dexdump) which is unsuitable for OSS anyways.

Agreed that it would make sense to split this into a separate PR though.

NotWoods replied:
Got it, thanks for breaking it down. Using the most popular FLOSS version instead makes sense to me.

gilbsgilbs replied:
#13767

NotWoods commented on app/build.gradle:
```diff
@@ -17,6 +17,8 @@ import org.mozilla.fenix.gradle.tasks.LintUnitTestRunner
 
 import static org.gradle.api.tasks.testing.TestResult.ResultType
 
+def useFreeImplementation = project.properties['org.mozilla.fenix.useFreeImplementation'].toBoolean()
```
This might be better off as another build variant, like debug.

gilbsgilbs replied:
So this is done. I've used build flavors which definitely makes more sense. Needing help to update taskcluster.

NotWoods commented on app/src/free/java/org/mozilla/fenix/components/metrics/AdvertisingIDImpl.kt:
```diff
@@ -0,0 +1,13 @@
+
+import android.content.Context
+
+interface AdvertisingIDImpl : AdvertisingID {
```
Should this be a class instead of an interface?

gilbsgilbs replied:
Rolledback to an interface because otherwise detekt was complaining of an abstract class containing only abstract method (see [UnncessaryAbstractClass](https://arturbosch.github.io/detekt/style.html#unnecessaryabstractclass)). Tell me if you think it would be better off keeping a class and ignoring the warning using an annotation.

NotWoods commented on app/src/main/java/org/mozilla/fenix/components/Push.kt:
```diff
@@ -42,5 +42,5 @@ class Push(context: Context, crashReporter: CrashReporter) {
     }
 
-    private val pushService by lazy { FirebasePushService() }
+    private val pushService by lazy { PushServiceImpl() }
```
Can we leave the class name as-is for readability?

gilbsgilbs commented:
I've moved to using build flavors instead of a build flag and fixed tests for both flavors. As expected, taskcluster will have to be updated to run the OSS tests suite and produce the OSS apks. I'd appreciate some help as I don't know how to test this locally.

gilbsgilbs commented:
I've tried to blindly update the taskcluster config, but it'll be hard to go further without being able to test. Please tell me there's anything I can do to get this PR to the next step. Thanks!

(and thoughts for all the Mozilla employees who were laid off :cry:)

pocmo commented:
Yeah, challenging times. I'll get back to that PR. Thank you so far for all that work! 👍

gilbsgilbs commented:
@pocmo thanks for your response. Although build flavors are generally the intended way to do this, I can totally understand your reasoning. I'll bring back build flags then.

pocmo commented:
Thank you! I am on PTO next week, so it may take a bit until can take another look.

gilbsgilbs commented:
Enjoy yourself :smile: .

gilbsgilbs commented:
I'll close this PR and make tinier, per-build flag PRs to make reviews easier.
